### PR TITLE
opentelemetry-cpp: init at 1.13.0

### PR DESCRIPTION
**Pull request by jfroche**
Build with support for:

- in memory
- otlp http
- otlp grpc
- zipkin
- prometheus
- elasticsearch

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  src = fetchFromGitHub {
+    owner = "open-telemetry";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-9dXA6DynYYoZno3Vz9SRP4o84EJa/XeiwB3pYaw+axU=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ cmake pkg-config protobuf abseil-cpp_202301 grpc ];
+  propagatedBuildInputs = [ gtest curl re2 prometheus-cpp nlohmann_json c-ares zlib ];
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:42]:
```suggestion
  buildInputs = [
    curl
    grpc
    gtest
    nlohmann_json
    prometheus-cpp
    protobuf
  ];
```

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  strictDeps = true;
+
+  cmakeFlags = [
+    "-DBUILD_SHARED_LIBS=ON"
+    "-DWITH_OTLP_HTTP=ON"
+    "-DWITH_OTLP_GRPC=ON"
+    "-DWITH_ABSEIL=ON"
+    "-DWITH_USE_NLOHMANN_JSON=ON"
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:50]: This is inferred.
```suggestion
```

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    "-DWITH_PROMETHEUS=ON"
+    "-DWITH_ELASTICSEARCH=ON"
+    "-DWITH_ZIPKIN=ON"
+    "-DOTELCPP_PROTO_PATH=${opentelemetry-proto}"
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:54]:
```suggestion
    "-DWITH_BENCHMARK=OFF"
    "-DOTELCPP_PROTO_PATH=${opentelemetry-proto}"
```

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+, abseil-cpp_202301
+, rsync
+, curl
+, grpc
+, c-ares
+, re2
+, prometheus-cpp
+, nlohmann_json
+, nix-update-script
+}:
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:19]: There is a bunch of unused stuff that you can get rid of.

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 5]
```diff
@@ -0,0 +1,67 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, gbenchmark
+, gtest
+, zlib
+, pkg-config
+, protobuf
+, abseil-cpp_202301
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:10]: Abseil is only needed for compatibility with old compilers and `protobuf`. The former is not necessary in current nixpkgs stdenvs and the latter propagates `abseil-cpp` itself.
```suggestion
```

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+let
+  opentelemetry-proto = fetchFromGitHub {
+    owner = "open-telemetry";
+    repo = "opentelemetry-proto";
+    rev = "v1.0.0";
+    sha256 = "sha256-1IylAZs8gElpruSX52A+ZopU8jXH/MjRE+FQV3gQ+Gk=";
+  };
+in
+stdenv.mkDerivation rec {
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:29]: Consider using `finalAttrs` instead of rec.

**Review by tobim — CHANGES_REQUESTED** [review]
Interesting library that should certainly be included in nixpkgs.

**Comment by jfroche** [comment]
> Interesting library that should certainly be included in nixpkgs.

@tobim thank you for your fast and thorough review !

**Comment by jfroche** [comment]
Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

**Comment by jfroche** [comment]
Result of `nixpkgs-review pr 262139` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

**Comment by jfroche** [comment]
Result of `nixpkgs-review pr 262139` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  outputs = [ "out" "dev" ];
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
+    description = "The OpenTelemetry C++ Client Library";
+    homepage = "https://github.com/open-telemetry/opentelemetry-cpp";
+    license = [ lib.licenses.asl20 ];
+    maintainers = with lib.maintainers; [ jfroche ];
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:68]:
```suggestion
    maintainers = with lib.maintainers; [ jfroche ];
    # https://github.com/protocolbuffers/protobuf/issues/14492
.   broken = !(stdenv.buildPlatform.canExecute stdenv.hostPlatform);
```

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 8]
```diff
@@ -0,0 +1,70 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, gtest
+, zlib
+, protobuf
+, curl
+, grpc
+, re2
```
Comment (tobim) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:10]:
```suggestion
, protobuf
, curl
, grpc
```

**Review by tobim — APPROVED** [review]
Looks good now apart from the now-unused params.

**Comment by panicgh** [comment]
You could enable most of the unit tests, see https://github.com/panicgh/nixpkgs/commit/d22827c52242e217b35b02bdf307dbc659056529.

I tried other ways to disable the tests than a patch, e.g. `GTEST_FILTER` or passing it to ctests (`ARGS="-E pattern"`), but it appears that they are always overwritten.

**Comment by jfroche** [comment]
> You could enable most of the unit tests, see [panicgh@d22827c](https://github.com/panicgh/nixpkgs/commit/d22827c52242e217b35b02bdf307dbc659056529).
> 
> I tried other ways to disable the tests than a patch, e.g. `GTEST_FILTER` or passing it to ctests (`ARGS="-E pattern"`), but it appears that they are always overwritten.

Thank you, I thought the tests were actually running.
I have included your change.

**Review by panicgh — APPROVED** [review]
LGTM.

Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

Edit: it may even be nice to backport it onto 23.11.

File: pkgs/by-name/op/opentelemetry-cpp/package.nix [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    sha256 = "sha256-1IylAZs8gElpruSX52A+ZopU8jXH/MjRE+FQV3gQ+Gk=";
+  };
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "opentelemetry-cpp";
+  version = "1.12.0";
+
+  src = fetchFromGitHub {
+    owner = "open-telemetry";
+    repo = finalAttrs.pname;
```
Comment (pbsds) [inline pkgs/by-name/op/opentelemetry-cpp/package.nix:28]:
```suggestion
    repo = "opentelemetry-cpp";
```

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

LGTM, just a quick nit

**Comment by jfroche** [comment]
Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

**Review by panicgh — APPROVED** [review]
Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

**Comment by pbsds** [comment]
It seems the `singleton.SingletonTest.Uniqueness` test fails on darwin on ofborg. Not sure what it does, but can it be reported upstream and disabled here in the meanwhile?

**Comment by panicgh** [comment]
According to ofborg, the `singleton.SingletonTest.Uniqueness` fails because of a segfault. I could not find the cause on a quick glance, but I agree that we can probably ignore it for the moment.

You may cherry-pick a fixup commit from https://github.com/panicgh/nixpkgs/commit/cd2847810c746295e363ddcbfd725864b0bb963d and squash it into your branch. It adds another patch to disable the `singleton.SingletonTest.Uniqueness` test.

I hope that we can get this package merged, finally.

**Comment by jfroche** [comment]
Thank you, almost there :crossed_fingers:

**Review by panicgh — APPROVED** [review]
Result of `nixpkgs-review pr 262139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

and it also seems to pass now on darwin.

**Pull request merged**